### PR TITLE
Iss#11

### PR DESCRIPTION
**Pull request by wallscheid**
Service pull request for review by WK

### CODE REVIEW

File: gym_electric_motor/visualization/console_printer.py [hunk 1]
```diff
@@ -1,23 +1,73 @@
...
-        print(
-            f'State {state * self._limits} \n'
-            f'Reference {reference * self._limits}\n'
-            f'Reward {reward}\n'
-            f'\n'
-        )
+    def step(self, k, state, reference, action, reward, done):
+        """Gets called at each step of the environment.
+        Handles per step printing as well es constraint violation printing"""
+        if self._print_freq:
```
Comment (wkirgsn) [inline gym_electric_motor/visualization/console_printer.py:46]: Misuse of integers for boolean expressions is no best practice.

```suggestion
        if self._print_freq > 0:
```

File: gym_electric_motor/visualization/console_printer.py [hunk 2]
```diff
@@ -1,23 +1,73 @@
...
+            if not self._num_steps:
```
Comment (wkirgsn) [inline gym_electric_motor/visualization/console_printer.py:47]:
```suggestion
            if self._num_steps == 0:
```

Please go over the whole document and correct the other occurences of treating integers as booleans
@XyDrKRulof

File: gym_electric_motor/visualization/motor_dashboard.py [hunk 3]
```diff
@@ -1,451 +1,118 @@
-from ..core import ElectricMotorVisualization
-from matplotlib import pyplot as plt
-import numpy as np
+from gym_electric_motor.core import ElectricMotorVisualization
+#import gym_electric_motor.visualization.motor_dashboard_plots as mdp
```
Comment (wkirgsn) [inline gym_electric_motor/visualization/motor_dashboard.py:2]: what is the use behind this commnt?

File: gym_electric_motor/visualization/motor_dashboard.py [hunk 4]
```diff
@@ -1,451 +1,118 @@
...
-        self.reference_points[k % self.update_cycle] = ref
-
-    def scatter(self, k):
+    def _initialize(self):
+        """Called with first render() call to setup the figures and plots.
         """
-        Update function for the dashboard. Plots all the points
+        plt.close()
+        assert len(self._plots)>0, "no plot variable selected"
+        # For the dark background lovers
```
Comment (wkirgsn) [inline gym_electric_motor/visualization/motor_dashboard.py:90]: awesome!

File: gym_electric_motor/visualization/motor_dashboard.py [hunk 5]
```diff
@@ -1,451 +1,118 @@
...
+            plt.style.use('dark_background')
+        self._figure, axes = plt.subplots(len(self._plots), sharex=True)
+        self._figure.subplots_adjust(wspace=0.0, hspace=0.2)
+        #plt.style.use("dark_background")
+        plt.xlabel('t/s')  # adding a common x-label to all the subplots
+
+        # if isinstance(axes, collections.Iterable):
+        #     print("list is iterable")
+
+        #plt.subplot() does not return an iterable var when the number of subplots==1
```
Comment (wkirgsn) [inline gym_electric_motor/visualization/motor_dashboard.py:101]: unnecessary comments

File: gym_electric_motor/visualization/motor_dashboard_plots.py [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+    def reset(self):
+        """Called by the MotorDashboard each time the environment is reset."""
+        pass
+
+
+class StatePlot(MotorDashboardPlot):
+    """Class to plot any motor state and its reference."""
+
+    # Width of the Plot in seconds
+    x_width = 3
```
Comment (wkirgsn) [inline gym_electric_motor/visualization/motor_dashboard_plots.py:55]: should this be adjustable by a user maybe?

File: gym_electric_motor/visualization/motor_dashboard_plots.py [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+        min_limit = self._limits * self._state_space[0] if self._normalized else self._state_space[0]
+        max_limit = self._limits * self._state_space[1] if self._normalized else self._state_space[1]
+
+        if self._state_space[0] < 0:
+            self._axis.axhline(min_limit, **self.limit_line_cfg)
+        lim = self._axis.axhline(max_limit, **self.limit_line_cfg)
+
+        self._axis.set_ylim(min_limit - 0.1 * (max_limit - min_limit), max_limit + 0.1*(max_limit - min_limit))
+        self._axis.set_xlim(0, self.x_width)
+        #self._axis.set_xlabel('t/s')
```
Comment (wkirgsn) [inline gym_electric_motor/visualization/motor_dashboard_plots.py:164]: orphaned comment

File: gym_electric_motor/visualization/motor_dashboard_plots.py [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+        self._axis.legend((base_action_line,), (self._action,), loc='upper left')
+
+    def set_modules(self, ps, rg, rf):
+        # fetch the action space from the physical system
+        self._action_space = ps.action_space
+        #extract the action index from the action name
+        self._action_idx = int(self._action.split('_')[1])
+        # check for the type of action space: Discrete or Continuous
+        if self._action_space.dtype == 'float32':  # for continuous action space
+            self._action_type = 'Continuous'
```
Comment (wkirgsn) [inline gym_electric_motor/visualization/motor_dashboard_plots.py:343]: self._mode goes with lower case "continuous" while self._action_type goes with CapitalCase "Continuous". Lets have everything in lower case

File: tests/test_visualizations/test_visualization.py [hunk 9]
```diff
@@ -7,19 +7,19 @@
...
-                                               ['omega', 'u', 'u_a', 'u_e']])
-def test_visualization(motor_type, converter_type, plotted_variables, turn_off_windows):
-    """
+#@pytest.mark.parametrize("motor_type", ['DcPermEx', 'DcShunt', 'DcExtEx'])
+#@pytest.mark.parametrize("converter_type", ['Disc-2QC', 'Disc-2QC', 'Disc-4QC', 'Cont-1QC', 'Cont-2QC', 'Cont-4QC'])
+#@pytest.mark.parametrize("plotted_variables", [['torque', 'omega', 'u_sup'], ['all'], ['none'], ['omega'],
+#                                             ['omega', 'u', 'u_a', 'u_e']])
+#def test_visualization(motor_type, converter_type, plotted_variables, turn_off_windows):
+"""
     test initialization and basic functions for all motor and converters
```
Comment (wkirgsn) [inline tests/test_visualizations/test_visualization.py:16]: What happened to this unit test?

File: gym_electric_motor/visualization/motor_dashboard_plots.py (hunk 7 above)
Comment (praneeth-b) [reply gym_electric_motor/visualization/motor_dashboard_plots.py:164]: fixed

File: gym_electric_motor/visualization/motor_dashboard.py (hunk 5 above)
Comment (praneeth-b) [reply gym_electric_motor/visualization/motor_dashboard.py:101]: fixed

File: gym_electric_motor/visualization/motor_dashboard.py (hunk 4 above)
Comment (praneeth-b) [reply gym_electric_motor/visualization/motor_dashboard.py:90]: fixed

File: gym_electric_motor/visualization/motor_dashboard.py (hunk 3 above)
Comment (praneeth-b) [reply gym_electric_motor/visualization/motor_dashboard.py:2]: removed unnecessary comments

File: gym_electric_motor/core.py [hunk 10]
```diff
@@ -291,9 +297,12 @@ def step(self, state, reference, reward, *_, **__):
             Only the first reference value is important here, because all others are not referenced.
 
         Args:
+            k(int): Current episode step
             state(ndarray(float)): The state of the physical system.
             reference(ndarray(float)): The reference array of the reference generator.
+            action(ndarray(float)): The last taken action. None after reset.
```
[the remark this answers is not in the document]
Comment (praneeth-b) [reply gym_electric_motor/core.py:303]: fixed

File: gym_electric_motor/core.py [hunk 11]
```diff
@@ -226,10 +231,11 @@ def step(self, action):
         """
         if self._reset_required:
             raise Exception('A reset is required before the environment can perform further steps')
```
[the remark this answers is not in the document]
Comment (praneeth-b) [reply gym_electric_motor/core.py:233]: fixed

File: examples/ddpg_pmsm_dq_current_control.py [hunk 12]
```diff
@@ -74,14 +74,14 @@ def reset(self, **kwargs):
...
-        visualization=MotorDashboard, visu_period=1,
+        visualization=MotorDashboard(['i_sq', 'i_sd', 'action_1', 'action_0']), visu_period=1,
         load=ConstantSpeedLoad(omega_fixed=1000 * np.pi / 30),
         control_space='dq',
         # Pass a string (with extra parameters)
         ode_solver='scipy.solve_ivp', solver_kwargs={},
         # Pass an instance
         reference_generator=rg,
-        plotted_variables=['i_sq', 'i_sd', 'u_sq', 'u_sd'],
+        #plotted_variables=['i_sq', 'i_sd', 'u_sq', 'u_sd', 'reward'],
```
[the remark this answers is not in the document]
Comment (praneeth-b) [reply examples/ddpg_pmsm_dq_current_control.py:84]: fixed

File: gym_electric_motor/visualization/console_printer.py (hunk 2 above)
Comment (XyDrKRulof) [reply gym_electric_motor/visualization/console_printer.py]: Done

**Comment by wallscheid** [comment]
Majaor refactoring of visualization successfully done incl review. Merge to nightly.

**Pull request merged**